CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxEnumFile.swift

```diff
@@ -57,4 +57,47 @@ let syntaxEnumFile = SourceFileSyntax(leadingTrivia: copyrightHeader) {
       }
     }
   }
+
+  for base: SyntaxNodeKind in [.decl, .stmt, .expr, .type, .pattern] {
+    let baseName = base.rawValue.withFirstCharacterUppercased;
+    try! EnumDeclSyntax(
+      """
+      /// Enum to exhaustively switch over all different \(raw: baseName) syntax nodes.
+      public enum \(raw: baseName)SyntaxEnum
+      """
+    ) {
+      for node in NON_BASE_SYNTAX_NODES where node.base == base {
+        DeclSyntax(
+          """
+          \(node.apiAttributes())\
+          case \(node.varOrCaseName)(\(node.kind.syntaxType))
+          """
+        )
+      }
+    }
+
+    try! ExtensionDeclSyntax(
+      """
+      public extension \(raw: baseName)Syntax
+      """
+    ) {
+      try FunctionDeclSyntax(
+        """
+        /// Get an enum that can be used to exhaustively switch over all \(raw: baseName) syntax nodes.
+        func `as`(_: \(raw: baseName)SyntaxEnum.Type) -> \(raw: baseName)SyntaxEnum
```

Comment:
How about adding an `is` method here as well? I believe it would be beneficial and consistent with the other casting methods we have.

Reply:
I don't think it's really valuable.  Is there any difference from e.g. `.is(ExprSyntax.self)`?

Reply:
Oh yeah, of course in this form it doesn't make any sense. What I had in mind – and didn't articulate well – is an `is` method that accepts a syntax enum type as an argument and returns a boolean indicating whether the node is of that type. It would look something like this:

```swift
extension ExprSyntax {
  func `is`(_ nodeType: ExprSyntaxEnum) -> Bool { ... }
}
```

What are your thoughts on this approach? While one could use the regular `is` method, I believe using an enum could make the API more user-friendly.

Reply:
You mean for exapple, like this?
```swift
func test(node1: ExprSyntax, node2: ExprSyntax) {
  let someExpr = ExprSyntaxEnum(node1)
  if node2.is(someExpr) {
    // node2 is the same type as node1
  }
}
```
Sorry, I personally still don't see a useful use-case of it. Maybe I'm still missing something.
Anyway, all the current `is(_:)` functions accept a meta type, not the value. Adding such method is outside the scope this PR.

Reply:
Since we already have `SyntaxEnum` with only an `as` method, I agree that expanding its scope might not fit this PR. However, while we're on the topic of these enums, I'd like to delve a bit deeper into whether this enhancement could be beneficial. Now that I've had my first coffee of the day, I hope to explain my idea more clearly 😅.

Currently, to check if a base node is of a specific leaf node type, we use the `is` method like this:
```swift
let someExpNode: ExprSyntax = "await completion?(argument)"

if someExpNode.is(AwaitExprSyntax.self) {
    // ...
}
```
With the new enums, we could potentially streamline this API:
```swift
let someExpNode: ExprSyntax = "await completion?(argument)"

if someExpNode.is(.awaitExpr) {
    // ...
}
```
The real question is, should we implement this?

I find it quite convenient to have an enumerated list of all possible node types to which a base node can be cast. It seems to make the API more intuitive and user-friendly. What do you think?

Reply:
I see. but `SyntaxEnum` and `XXXSyntaxEnum` cases have the associated value. So `is(.awaitExpr)` is not possible using these `*SyntaxEnum` types.
The question would be, should we make `{Decl|Stmt|Expr|Type|Pattern}SyntaxKind`  and ``func `is`(_ kind: ExprSyntaxKind) -> Bool`` is valuable and not confusing while we already have `node.kind == .awaitExpr`. I recommend to open an issue, and continue discussion there with wider audience.